### ISSUE TITLE
Initial prototype

### ISSUE THREAD

**Issue by simonw** [simonw/datasette-granian#1]
Idea is to use the same pattern as https://github.com/simonw/datasette-gunicorn but serve Datasette using https://github.com/emmett-framework/granian instead.

**Comment by simonw** [comment]
This doesn't work. I get this error:

```
% datasette granian .                                                                     
[INFO] Starting granian
[INFO] Listening at: 127.0.0.1:8001
Traceback (most recent call last):
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/bin/datasette", line 8, in <module>
    sys.exit(cli())
             ^^^^^
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/simon/Dropbox/Development/datasette-granian/datasette_granian/__init__.py", line 47, in serve_with_granian
    ).serve()
      ^^^^^^^
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/granian/server.py", line 312, in serve
    self.startup(spawn_target, partial(target_loader, self.target))
  File "/Users/simon/.local/share/virtualenvs/datasette-granian-tlvX8pqZ/lib/python3.11/site-packages/granian/server.py", line 292, in startup
    proc.start()
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/context.py", line 288, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 32, in __init__
    super().__init__(process_obj)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/popen_fork.py", line 19, in __init__
    self._launch(process_obj)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/popen_spawn_posix.py", line 47, in _launch
    reduction.dump(process_obj, fp)
  File "/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/multiprocessing/reduction.py", line 60, in dump
    ForkingPickler(file, protocol).dump(obj)
AttributeError: Can't pickle local object 'asgi_csrf_decorator.<locals>._asgi_csrf_decorator.<locals>.app_wrapped_with_csrf'
```

**Comment by simonw** [comment]
Wrote this code to replicate the bug:
```python
from granian.server import Granian
from granian.constants import HTTPModes, Interfaces, Loops, ThreadModes
from granian.log import LogLevels
from datasette.app import Datasette


def serve_with_granian():
    ds = Datasette(memory=True)
    app = ds.app()
    Granian(
        app,
        address="127.0.0.1",
        port=8002,
        interface=Interfaces.ASGI,
        workers=1,
        threads=1,
        pthreads=1,
        threading_mode=ThreadModes.workers.value,
        loop=Loops.auto.value,
        http=HTTPModes.auto.value,
        websockets=True,
        backlog=1024,
        log_level=LogLevels.info.value,
        ssl_cert=None,
        ssl_key=None,
    ).serve()


if __name__ == "__main__":
    serve_with_granian()
```

**Comment by simonw** [comment]
Filed an issue:
- https://github.com/emmett-framework/granian/issues/35

**Comment by simonw** [comment]
I just tested this against the newly released Granian 1.0 and it still works: https://github.com/emmett-framework/granian/releases/tag/v1.0.0
